File: src/main/java/com/chrisrm/idea/themes/models/MTBundledTheme.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2019 Chris Magnussen and Elior Boukhobza
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 *
 */

package com.chrisrm.idea.themes.models;

import com.chrisrm.idea.themes.models.parsers.MTBundledThemeParser;
import com.chrisrm.idea.themes.themes.MTAbstractTheme;
import com.intellij.util.ObjectUtils;
import com.intellij.util.xmlb.annotations.Property;
import com.intellij.util.xmlb.annotations.Tag;
import org.jetbrains.annotations.NotNull;

import javax.swing.plaf.ColorUIResource;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public abstract class MTBundledTheme extends MTAbstractTheme {
  @Tag
  private final List<MTThemeColor> colors = new ArrayList<>(16);
  @Property
  protected String themeId;

  @Property
  @Deprecated
  protected String accentColor;
  @Property
  @Deprecated
  protected String excludedColor;

  /**
   * The theme parser, according to the bridge design pattern every subclass must define the parser
   */
  protected abstract MTBundledThemeParser getThemeParser();

  @SuppressWarnings("NoopMethodInAbstractClass")
  @Override
  protected void init() {
  }

  @Override
  public final String getThemeName() {
    return getName();
  }

  //region --------------- Setters ------------------
  public final void setExcludedColor(final ColorUIResource excludedColor) {
    getThemeParser().setExcludedColor(excludedColor);
  }

  public final void setAccentColor(final ColorUIResource accentColor) {
    getThemeParser().setAccentColor(accentColor);
  }

  public final void setNotificationsColor(final ColorUIResource notificationsColor) {
    getThemeParser().setNotificationsColor(notificationsColor);
  }

  public final void setHighlightColor(final ColorUIResource highlightColor) {
    getThemeParser().setHighlightColor(highlightColor);
  }

  public final void setTreeSelectionColor(final ColorUIResource treeSelectionColor) {
    getThemeParser().setTreeSelectionColor(treeSelectionColor);
  }

  public final void setSecondBorderColor(final ColorUIResource secondBorderColor) {
    getThemeParser().setSecondBorderColor(secondBorderColor);
  }

  public final void setTableSelectedColor(final ColorUIResource tableSelectedColor) {
    getThemeParser().setTableSelectedColor(tableSelectedColor);
  }

  public final void setContrastColor(final ColorUIResource contrastColor) {
    getThemeParser().setContrastColor(contrastColor);
  }

  public final void setDisabledColor(final ColorUIResource disabledColor) {
    getThemeParser().setDisabledColor(disabledColor);
  }

  public final void setSecondaryBackgroundColor(final ColorUIResource secondaryBackgroundColor) {
    getThemeParser().setSecondaryBackgroundColor(secondaryBackgroundColor);
  }

  public final void setButtonColor(final ColorUIResource buttonColor) {
    getThemeParser().setButtonColor(buttonColor);
  }

  public final void setSelectionForegroundColor(final ColorUIResource selectionForegroundColor) {
    getThemeParser().setSelectionForegroundColor(selectionForegroundColor);
  }

  public final void setSelectionBackgroundColor(final ColorUIResource selectionBackgroundColor) {
    getThemeParser().setSelectionBackgroundColor(selectionBackgroundColor);
  }

  public final void setTextColor(final ColorUIResource textColor) {
    getThemeParser().setTextColor(textColor);
  }

  public final void setForegroundColor(final ColorUIResource foregroundColor) {
    getThemeParser().setForegroundColor(foregroundColor);
  }

  public final void setBackgroundColor(final ColorUIResource backgroundColor) {
    getThemeParser().setBackgroundColor(backgroundColor);
  }

  public final void setThemeId(final String themeId) {
    this.themeId = themeId;
  }
  //endregion

  //region --------------- Getters ------------------

  /**
   * Get and parse the accent color
   */
  @Override
  public final ColorUIResource getAccentColorResource() {
    return getThemeParser().getAccentColorString();
  }

  /**
   * Get and parse the excluded color
   */
  @Override
  public final ColorUIResource getExcludedColorResource() {
    return getThemeParser().getExcludedColorString();
  }

  @Override
  public final ColorUIResource getBackgroundColorResource() {
    return getThemeParser().getBackgroundColorString();
  }

  @Override
  public final ColorUIResource getForegroundColorResource() {
    return getThemeParser().getForegroundColorString();
  }

  @Override
  public final ColorUIResource getTextColorResource() {
    return getThemeParser().getTextColorString();
  }

  @Override
  public final ColorUIResource getSelectionBackgroundColorResource() {
    return getThemeParser().getSelectionBackgroundColorString();
  }

  @Override
  public final ColorUIResource getSelectionForegroundColorResource() {
    return getThemeParser().getSelectionForegroundColorString();
  }

  @Override
  public final ColorUIResource getButtonColorResource() {
    return getThemeParser().getButtonColorString();
  }

  @Override
  public final ColorUIResource getSecondaryBackgroundColorResource() {
    return getThemeParser().getSecondaryBackgroundColorString();
  }

  @Override
  public final ColorUIResource getDisabledColorResource() {
    return getThemeParser().getDisabledColorString();
  }

  @Override
  public final ColorUIResource getContrastColorResource() {
    return getThemeParser().getContrastColorString();
  }

  @Override
  public final ColorUIResource getTableSelectedColorResource() {
    return getThemeParser().getTableSelectedColorString();
  }

  @Override
  public final ColorUIResource getSecondBorderColorResource() {
    return getThemeParser().getSecondBorderColorString();
  }

  @Override
  public final ColorUIResource getHighlightColorResource() {
    return getThemeParser().getHighlightColorString();
  }

  @Override
  public final ColorUIResource getTreeSelectionColorResource() {
    return getThemeParser().getTreeSelectionColorString();
  }

  @Override
  public final ColorUIResource getNotificationsColorResource() {
    return getThemeParser().getNotificationsColorString();
  }

  @NotNull
  @Override
  public final String getThemeId() {
    return ObjectUtils.notNull(themeId, "");
  }

  /**
   * Add a big number for the order so it appears at the end
   */
  @Override
  public abstract int getOrder();

  public final Collection<MTThemeColor> getColors() {
    return colors;
  }
  //endregion
}

File: src/main/java/com/mallowigi/idea/themes/models/MTBundledTheme.kt
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015-2021 Elior "Mallowigi" Boukhobza
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 *
 */
package com.mallowigi.idea.themes.models

import com.intellij.util.xmlb.annotations.Tag
import com.mallowigi.idea.themes.MTBundledThemesManager
import com.mallowigi.idea.themes.models.parsers.MTBundledThemeParser
import com.mallowigi.idea.themes.themes.MTAbstractTheme
import com.thoughtworks.xstream.annotations.XStreamAlias
import com.thoughtworks.xstream.annotations.XStreamOmitField
import org.jetbrains.annotations.NonNls
import javax.swing.plaf.ColorUIResource

/**
 * Model for External and Custom Themes
 *
 */
@XStreamAlias("mtTheme")
abstract class MTBundledTheme : MTAbstractTheme() {

  @XStreamAlias("themeId")
  @NonNls
  override var themeId: String = ""

  /**
   * Theme ID
   * @Deprecated use [.themeId] instead
   */
  @XStreamAlias("id")
  @XStreamOmitField
  var id: String = ""

  @XStreamAlias("name")
  @NonNls
  override var themeName: String = ""

  @XStreamAlias("editorColorsScheme")
  @NonNls
  override var themeColorScheme: String? = null

  /**
   * Is theme dark?
   */
  @Suppress("HardCodedStringLiteral")
  @XStreamAlias("dark")
  override var isThemeDark: Boolean = true

  /**
   * Theme Colors
   */
  @Suppress("HardCodedStringLiteral")
  @Tag
  @XStreamAlias("colors")
  val colors: List<MTThemeColor> = ArrayList(16)

  /**
   * The theme parser, according to the bridge design pattern every subclass must define the parser
   */
  protected abstract val themeParser: MTBundledThemeParser

  //region --------------- Getters ------------------

  override val backgroundColorResource: ColorUIResource
    get() = themeParser.backgroundColorString

  override val foregroundColorResource: ColorUIResource
    get() = themeParser.foregroundColorString

  override val textColorResource: ColorUIResource
    get() = themeParser.textColorString

  override val selectionBackgroundColorResource: ColorUIResource
    get() = themeParser.selectionBackgroundColorString

  override val selectionForegroundColorResource: ColorUIResource
    get() = themeParser.selectionForegroundColorString

  override val buttonColorResource: ColorUIResource
    get() = themeParser.buttonColorString

  override val secondaryBackgroundColorResource: ColorUIResource
    get() = themeParser.secondaryBackgroundColorString

  override val disabledColorResource: ColorUIResource
    get() = themeParser.disabledColorString

  override val contrastColorResource: ColorUIResource
    get() = themeParser.contrastColorString

  override val tableSelectedColorResource: ColorUIResource
    get() = themeParser.tableSelectedColorString

  override val secondBorderColorResource: ColorUIResource
    get() = themeParser.secondBorderColorString

  override val highlightColorResource: ColorUIResource
    get() = themeParser.highlightColorString

  override val treeSelectionColorResource: ColorUIResource
    get() = themeParser.treeSelectionColorString

  override val notificationsColorResource: ColorUIResource
    get() = themeParser.notificationsColorString

  override val accentColorResource: ColorUIResource
    get() = themeParser.accentColorString

  override val excludedColorResource: ColorUIResource
    get() = themeParser.excludedColorString

  abstract override val order: Int

  override val backgroundImage: String?
    get() = null
  //endregion

  //region --------------- Setters ------------------
  /**
   * Excluded Color
   */
  fun setExcludedColor(excludedColor: ColorUIResource): Unit = themeParser.setExcludedColor(excludedColor)

  /**
   * Accent Color
   */
  fun setAccentColor(accentColor: ColorUIResource): Unit = themeParser.setAccentColor(accentColor)

  /**
   * Notifications Color
   */
  fun setNotificationsColor(notificationsColor: ColorUIResource): Unit =
    themeParser.setNotificationsColor(notificationsColor)

  /**
   * Highlight Color
   */
  fun setHighlightColor(highlightColor: ColorUIResource): Unit = themeParser.setHighlightColor(highlightColor)

  /**
   * Tree Selection Color
   */
  fun setTreeSelectionColor(treeSelectionColor: ColorUIResource): Unit =
    themeParser.setTreeSelectionColor(treeSelectionColor)

  /**
   * Second Border Color
   */
  fun setSecondBorderColor(secondBorderColor: ColorUIResource): Unit =
    themeParser.setSecondBorderColor(secondBorderColor)

  /**
   * Table Selected Color
   */
  fun setTableSelectedColor(tableSelectedColor: ColorUIResource): Unit =
    themeParser.setTableSelectedColor(tableSelectedColor)

  /**
   * Contrast Color
   */
  fun setContrastColor(contrastColor: ColorUIResource): Unit = themeParser.setContrastColor(contrastColor)

  /**
   * Disabled Color
   */
  fun setDisabledColor(disabledColor: ColorUIResource): Unit = themeParser.setDisabledColor(disabledColor)

  /**
   * Secondary Background Color
   */
  fun setSecondaryBackgroundColor(secondaryBackgroundColor: ColorUIResource): Unit =
    themeParser.setSecondaryBackgroundColor(secondaryBackgroundColor)

  /**
   * Button Color
   */
  fun setButtonColor(buttonColor: ColorUIResource): Unit = themeParser.setButtonColor(buttonColor)

  /**
   * Selection Foreground Color
   */
  fun setSelectionForegroundColor(selectionForegroundColor: ColorUIResource): Unit =
    themeParser.setSelectionForegroundColor(selectionForegroundColor)

  /**
   * Selection Background Color
   */
  fun setSelectionBackgroundColor(selectionBackgroundColor: ColorUIResource): Unit =
    themeParser.setSelectionBackgroundColor(selectionBackgroundColor)

  /**
   * Text Color
   */
  fun setTextColor(textColor: ColorUIResource): Unit = themeParser.setTextColor(textColor)

  /**
   * Foregroun Color
   */
  fun setForegroundColor(foregroundColor: ColorUIResource): Unit = themeParser.setForegroundColor(foregroundColor)

  /**
   * Background Color
   */
  fun setBackgroundColor(backgroundColor: ColorUIResource): Unit = themeParser.setBackgroundColor(backgroundColor)

  //endregion

  /**
   * Save theme from the form data to a file
   */
  fun saveTheme(name: String, id: String, dark: Boolean, colorScheme: String?) {
    this.themeName = name
    this.themeId = id
    this.themeColorScheme = colorScheme
    this.isThemeDark = dark

    MTBundledThemesManager.saveTheme(this)
  }
}
